### PR TITLE
Add local development capability to the pipeline repo

### PR DESCRIPTION
**Pull request by marcellp**
# Background
#### Link to issue 
https://biomage.atlassian.net/browse/BIOMAGE-508

#### Link to staging deployment URL 
N/A

#### Links to any Pull Requests related to this
N/A

#### Anything else the reviewers should know about the changes here

# Changes
### Code changes
Added a new project, `local-runner`, that can automatically capture and display logs by containers that are run using AWS Step Functions through InfraMock.

Additionally, added a new CloudFormation file that will create two Lambda functions. Both are deployed locally to InfraMock, and they allow us to run Docker containers without relying on Kubernetes. The API will use these lambda functions instead of Kubernetes Pods when it's running in development to construct the pipeline.

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [x] Tested locally with Inframock
- [ ] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [x] Relevant Github READMEs updated
- [x] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [x] Photo of a cute animal attached to this PR
image

### CODE REVIEW

File: local-runner/cf-local-container-launcher.yaml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+          def random_string(length):
+            return ''.join(random.choice(string.ascii_letters) for m in range(length))
+
+          def handler(event, context):
+            proc = subprocess.run(f"docker run --rm --name {event['name']}-{random_string(10)} -d {event['image']}", shell=True, check=True)
+
+            return proc.returncode
+      Runtime: "python3.8"
+      Timeout: "25"
```
Comment (marcellp) [inline local-runner/cf-local-container-launcher.yaml:34]: A mock function. Its only purpose is to launch a given container with a custom unique name such as `pipeline-remoter-server-xmNolHtobr` and a custom image. It is running in detached mode (`-d`) and the container is destroyed once it quits (`--rm`). This mimicks the behavior of how Pods work in Kubernetes.

File: local-runner/cf-local-container-launcher.yaml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        ZipFile: |
+          import subprocess
+
+          # Removes all containers that already exist and contain 'pipeline' in their name.
+          def handler(event, context):
+            proc = subprocess.run("docker kill $(docker ps -f name=pipeline --format '{{.Names}}') || true", shell=True, check=True)
+
+            return proc.returncode
+      Runtime: "python3.8"
+      Timeout: "25"
```
Comment (marcellp) [inline local-runner/cf-local-container-launcher.yaml:52]: Another Lambda function, which is equivalent to the step in production/staging where used/completed jobs are destroyed. Here, all containers with `pipeline` in their name are killed.

**Comment by marcellp** [comment]
Note: you can test the functionality of the script manually, but this will not really work properly until the API changes for local testing are deployed as well.

**Review by cosa65 — COMMENTED** [review]
Looks good, we should update the readme with the ecr login pre requirement.

**Review by cosa65 — APPROVED** [review]

**Pull request merged**